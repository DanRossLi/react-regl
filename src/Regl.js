import React, { Component } from 'react'
import ReactDOM from 'react-dom';
import ReglInit from 'regl';

import ReactUpdates from 'react-dom/lib/ReactUpdates';
import ReactInstanceMap from 'react-dom/lib/ReactInstanceMap';

import ContainerMixin from './ContainerMixin';

import Node from 'display-tree';

import batchChildren from './util/batchChildren';
import topDownDrawScopes from './util/topDownDrawScopes';

import PropTypes from 'prop-types';

class Regl extends Component {
  constructor(props, context){
    super(props, context);
    
    this.state = {
      width: props.width || window.innerWidth,
      height: props.height || window.innerHeight
    };
  }

  
  componentDidMount() {
    this._debugID = this._reactInternalInstance._debugID;

    this._hostContainerInfo = {
      _idCounter: 0
    };

    this.node = Node();
    this.node.type = "Regl";

    const canvasRef = this.props.canvas || this.refs.canvas;

    const regl = ReglInit(canvasRef);
    regl.renderers = {};
    this.regl = regl;

    if(this.props.onFrame){
      regl.frame(this.props.onFrame);
    }

    const transaction = ReactUpdates.ReactReconcileTransaction.getPooled();
    transaction.perform(
      this.mountAndInjectChildren,
      this,
      this.props.children,
      transaction,
      Object.assign({}, ReactInstanceMap.get(this)._context, { regl })
    );
    
    ReactUpdates.ReactReconcileTransaction.release(transaction);

    this.drawScope = topDownDrawScopes(this.node);
    this.drawScope();
  }
  
  componentDidUpdate(oldProps) {
    const transaction = ReactUpdates.ReactReconcileTransaction.getPooled();
    transaction.perform(
      this.updateChildren,
      this,
      this.props.children,
      transaction,
      Object.assign({}, ReactInstanceMap.get(this)._context, { regl: this.regl })
    );
    ReactUpdates.ReactReconcileTransaction.release(transaction);


    if(this.props.clear){
      this.regl.clear(this.props.clear);
    }

    this.drawScope();
  }
  
  render() {
    if(this.props.canvas){
      return null;
    }
    
    const { width, height } = this.state;
    return (
      <canvas ref="canvas" width={width} height={height} />
    )
  }
}

Object.assign(Regl.prototype, ContainerMixin);

Regl.childContextTypes = {
  regl: PropTypes.func
};

export default Regl;
