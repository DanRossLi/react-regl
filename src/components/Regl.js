import React from 'react';
import PropTypes from 'prop-types';
import ReglRenderer from '../renderer.js';
import reglInit from 'regl';

import ReglRootNode from '../nodes/ReglRootNode';

export default class Regl extends React.Component {
  regl = null
  tick = null;
  canvasRef = null;
  rootNode = null;

  componentWillUnmount(){
    ReglRenderer.updateContainer(null, this.mountNode, this);
    this.regl.destroy();
  }

  constructor(props, context){
    super(props, context);
  }

  getChildContext(){
    if(this.context && this.context.store){
      return {store: this.context.store};
    }
    return this.context;
  }

  componentDidMount(){
    const canvasRef = this.props.canvas || this.canvasRef;
    const gl = canvasRef.getContext("webgl", {
      alpha: false,
      antialias: false,
      stencil: false,
      preserveDrawingBuffer: false
    });

    const regl = reglInit({ gl });
    regl.cache = {};
    this.regl = regl;

    if(this.props.onReglInit &&
       typeof this.props.onReglInit === 'function'){
      this.props.onReglInit(regl);
    }

    this.rootNode = ReglRenderer.createContainer(new ReglRootNode(regl, this.context));

    this.regl.clear({
      color: this.props.color || [0, 0, 0, 1],
      depth: this.props.depth || 1
    });

    ReglRenderer.updateContainer(this.props.children, this.rootNode, this);
    this.rootNode.containerInfo.render();

    if(this.props.forceRedrawOnTick === true){
      this.tick = regl.frame(() => {
        this.regl.clear({
          color: this.props.color || [0, 0, 0, 1],
          depth: this.props.depth || 1
        });
        this.rootNode.containerInfo.render();
      });
    }

    if(this.props.onFrame && typeof this.props.onFrame === 'function'){
      this.tick = regl.frame(this.props.onFrame);
    }
  }

  componentDidUpdate(prevProps, prevState){
    this.regl.clear({
      color: this.props.color || [0, 0, 0, 1],
      depth: this.props.depth || 1
    });

    if(prevProps.width !== this.props.width ||
       prevProps.height !== this.props.height){
      this.regl.poll();
    }

    ReglRenderer.updateContainer(this.props.children, this.rootNode, this);
    this.rootNode.containerInfo.render();
  }

  componentWillUnmount(){
    if(this.tick){
      this.tick.cancel();
    }

    if(this.regl){
      this.regl.destroy();
    }
  }

  render(){
    if(this.props.canvas){
      return null;
    }

    return (
      <canvas ref={inst => {this.canvasRef = inst}}
              width={this.props.width}
              height={this.props.height} />
    )
  }
}
// In order to support react-redux apps
// the store needs to be passed to the reconciler on context
Regl.contextTypes = { store: PropTypes.object };
Regl.childContextTypes = { store: PropTypes.object };
