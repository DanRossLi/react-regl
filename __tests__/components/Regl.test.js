import React from 'react';
import Regl from '../../src/index.js';
import ReactTestRenderer from 'react-test-renderer';
import gl from 'gl';

const mockCanvas = {
  getContext: () => {
     return gl(800, 600, {
      alpha: false,
      antialias: false,
      stencil: false,
      preserveDrawingBuffer: false
    });
  }
}

it('should render a canvas', () => {
  /* expect(ReactTestRenderer.create(
   *   <Regl canvas={mockCanvas}
   *         width={500}
   *         height={400} />
   * )).toMatchSnapshot();*/
})
