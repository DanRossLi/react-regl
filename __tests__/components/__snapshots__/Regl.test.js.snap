// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a canvas 1`] = `null`;
